File: content/tracks/cpp/concepts/09-concurrency/concurrency_test.cpp
// Compile: g++ -std=c++17 -pthread -o test concurrency_test.cpp && ./test
#include <atomic>
#include <thread>
#include <mutex>
#include <future>
#include <vector>
#include <numeric>
#include <cstdio>

class Counter {
public:
    Counter() : value_(0) {}
    void increment() { value_++; }
    void decrement() { value_--; }
    int value() const { return value_.load(); }
private:
    std::atomic<int> value_;
};

int parallel_sum(const std::vector<int>& data, int num_threads);
std::future<int> async_square(int n);

static int passed = 0, failed = 0;
#define CHECK(name, cond) \
    do { if (cond) { printf("  PASS: %s\n", name); passed++; } \
         else      { printf("  FAIL: %s\n", name); failed++; } } while(0)

int main() {
    // Counter
    {
        Counter c;
        CHECK("Counter: initial value 0", c.value() == 0);
        c.increment(); c.increment();
        CHECK("Counter: after 2 increments", c.value() == 2);
        c.decrement();
        CHECK("Counter: after decrement", c.value() == 1);
    }

    // Counter: concurrent increments
    {
        Counter c;
        std::vector<std::thread> threads;
        for (int i = 0; i < 100; ++i)
            threads.emplace_back([&] { c.increment(); });
        for (auto& t : threads) t.join();
        CHECK("Counter: 100 concurrent increments", c.value() == 100);
    }

    // parallel_sum
    {
        std::vector<int> data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        CHECK("parallel_sum: 1 thread",  parallel_sum(data, 1)  == 55);
        CHECK("parallel_sum: 2 threads", parallel_sum(data, 2)  == 55);
        CHECK("parallel_sum: 4 threads", parallel_sum(data, 4)  == 55);
        CHECK("parallel_sum: empty",     parallel_sum({}, 4)    == 0);
        CHECK("parallel_sum: 1 element", parallel_sum({42}, 2)  == 42);
    }

    // async_square
    {
        auto f1 = async_square(5);
        CHECK("async_square: 5^2 = 25", f1.get() == 25);
        auto f2 = async_square(0);
        CHECK("async_square: 0^2 = 0", f2.get() == 0);
        auto f3 = async_square(-3);
        CHECK("async_square: (-3)^2 = 9", f3.get() == 9);
    }

    printf("\n%d passed, %d failed\n", passed, failed);
    return failed > 0 ? 1 : 0;
}

File: content/tracks/rust/concepts/10-concurrency/concurrency_test.rs
// Compile and run: rustc concurrency_test.rs -o concurrency_test && ./concurrency_test

mod solution {
    use std::sync::{Arc, Mutex, mpsc};
    use std::thread;

    pub fn parallel_sum(n: u32) -> i32 {
        let counter = Arc::new(Mutex::new(0i32));
        let handles: Vec<_> = (0..n)
            .map(|i| {
                let c = Arc::clone(&counter);
                thread::spawn(move || {
                    let mut num = c.lock().unwrap();
                    *num += i as i32;
                })
            })
            .collect();
        for h in handles {
            h.join().unwrap();
        }
        *counter.lock().unwrap()
    }

    pub fn parallel_double(items: Vec<i32>) -> Vec<i32> {
        let (tx, rx) = mpsc::channel();
        let handles: Vec<_> = items
            .into_iter()
            .map(|item| {
                let tx = tx.clone();
                thread::spawn(move || {
                    tx.send(item * 2).unwrap();
                })
            })
            .collect();
        drop(tx);
        for h in handles {
            h.join().unwrap();
        }
        rx.iter().collect()
    }
}

fn main() {
    let mut passed = 0u32;
    let mut failed = 0u32;

    macro_rules! check {
        ($name:expr, $cond:expr) => {
            if $cond {
                println!("  PASS: {}", $name);
                passed += 1;
            } else {
                println!("  FAIL: {}", $name);
                failed += 1;
            }
        };
    }

    // parallel_sum: 0+1+2+3+4 = 10
    check!("parallel_sum(5) = 10", solution::parallel_sum(5) == 10);
    // 0 threads: identity is 0
    check!("parallel_sum(0) = 0", solution::parallel_sum(0) == 0);
    // 1 thread: only index 0 -> 0
    check!("parallel_sum(1) = 0", solution::parallel_sum(1) == 0);
    // 0+1+...+9 = 45
    check!("parallel_sum(10) = 45", solution::parallel_sum(10) == 45);

    // parallel_double
    let mut r1 = solution::parallel_double(vec![1, 2, 3, 4]);
    r1.sort();
    check!("parallel_double [1,2,3,4]", r1 == vec![2, 4, 6, 8]);

    let r2 = solution::parallel_double(vec![]);
    check!("parallel_double empty", r2.is_empty());

    let mut r3 = solution::parallel_double(vec![0, -1, 5]);
    r3.sort();
    check!("parallel_double [0,-1,5]", r3 == vec![-2, 0, 10]);

    let mut r4 = solution::parallel_double(vec![100]);
    r4.sort();
    check!("parallel_double single element", r4 == vec![200]);

    println!("\n{} passed, {} failed", passed, failed);
    if failed > 0 {
        std::process::exit(1);
    }
}
